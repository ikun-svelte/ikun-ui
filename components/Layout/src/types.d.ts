/// <reference types="svelte" />
import type { ClassValue } from 'clsx';
export type KLayoutProps = {
	cls: ClassValue;
	attrs: Record<string, string>;
};
export type KHeaderFooterProps = {
	height: string;
};
export type KAsideProps = {
	width: string;
};
export type KContainerProps = {
	direction: 'vertical' | 'horizontal';
};
