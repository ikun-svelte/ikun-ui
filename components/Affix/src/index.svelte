<script lang="ts">
	import { getPrefixCls } from '@ikun-ui/utils';
	import { clsx } from 'clsx';
	import type { KAffixProps } from './types';
	import type { ScrollTarget } from './utils';
	import { unwrapElement, getScrollTop, getRect, beforeNextFrameOnce } from './utils';
	import { onDestroy, onMount } from 'svelte';
	export let cls: KAffixProps['cls'] = undefined;
	export let attrs: KAffixProps['attrs'] = {};
	export let listenTo: KAffixProps['listenTo'] = undefined;
	export let top: KAffixProps['top'] = undefined;
	export let bottom: KAffixProps['bottom'] = undefined;
	export let triggerTop: KAffixProps['triggerTop'] = undefined;
	export let triggerBottom: KAffixProps['triggerBottom'] = undefined;
	export let positionOption: KAffixProps['positionOption'] = 'fixed';

	let scrollTarget: ScrollTarget | null = null;
	let stickToTopRef = false;
	let stickToBottomRef = false;
	let bottomAffixedTriggerScrollTopRef: number | null = null;
	let topAffixedTriggerScrollTopRef: number | null = null;
	$: affixedRef = stickToBottomRef || stickToTopRef;
	$: mergedOffsetTopRef = triggerTop ?? top;
	$: mergedTopRef = top ?? triggerTop;
	$: mergedBottomRef = bottom ?? triggerBottom;
	$: mergedOffsetBottomRef = triggerBottom ?? bottom;
	let selfRef: Element | null = null;

	const init = (): void => {
		if (listenTo) {
			scrollTarget = unwrapElement(listenTo);
		} else {
			scrollTarget = document;
		}
		if (scrollTarget) {
			scrollTarget.addEventListener('scroll', handleScroll);
			handleScroll();
		}
	};
	function handleScroll(): void {
		beforeNextFrameOnce(doHandleScroll);
	}

	function doHandleScroll(): void {
		const selfEl = selfRef;
		if (!scrollTarget || !selfEl) return;
		const scrollTop = getScrollTop(scrollTarget);
		if (affixedRef) {
			if (topAffixedTriggerScrollTopRef !== null && scrollTop < topAffixedTriggerScrollTopRef) {
				stickToTopRef = false;
				topAffixedTriggerScrollTopRef = null;
			}
			if (
				bottomAffixedTriggerScrollTopRef !== null &&
				scrollTop > bottomAffixedTriggerScrollTopRef
			) {
				stickToBottomRef = false;
				bottomAffixedTriggerScrollTopRef = null;
			}
			return;
		}
		const containerRect = getRect(scrollTarget);
		const affixRect = selfEl.getBoundingClientRect();
		const pxToTop = affixRect.top - containerRect.top;
		const pxToBottom = containerRect.bottom - affixRect.bottom;
		const mergedOffsetTop = mergedOffsetTopRef;
		const mergedOffsetBottom = mergedOffsetBottomRef;
		if (mergedOffsetTop !== undefined && pxToTop <= mergedOffsetTop) {
			stickToTopRef = true;
			topAffixedTriggerScrollTopRef = scrollTop - (mergedOffsetTop - pxToTop);
		} else {
			stickToTopRef = false;
			topAffixedTriggerScrollTopRef = null;
		}
		if (mergedOffsetBottom !== undefined && pxToBottom <= mergedOffsetBottom) {
			stickToBottomRef = true;
			bottomAffixedTriggerScrollTopRef = scrollTop + mergedOffsetBottom - pxToBottom;
		} else {
			stickToBottomRef = false;
			bottomAffixedTriggerScrollTopRef = null;
		}
	}
	onMount(init);
	onDestroy(() => {
		if (!scrollTarget) return;
		scrollTarget.removeEventListener('scroll', handleScroll);
	});

	let styleTop: string = '';
	$: {
		if (stickToTopRef && mergedOffsetTopRef !== undefined && mergedTopRef !== undefined) {
			styleTop = `${mergedTopRef}px`;
		}
	}

	let styleBottom: string = '';
	$: {
		if (stickToBottomRef && mergedOffsetBottomRef !== undefined && mergedBottomRef !== undefined) {
			styleBottom = `${mergedBottomRef}px`;
		}
	}

	const prefixCls = getPrefixCls('affix');
	$: cnames = clsx(
		prefixCls,
		{
			[`${prefixCls}--affixed`]: affixedRef,
			[`${prefixCls}--absolute-positioned`]: positionOption === 'absolute'
		},
		cls
	);
</script>

<div
	class={cnames}
	style:top={styleTop}
	style:bottom={styleBottom}
	bind:this={selfRef}
	{...$$restProps}
	{...attrs}
>
	<slot />
</div>
