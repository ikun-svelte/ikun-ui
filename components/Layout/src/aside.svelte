<script lang="ts">
	import { getPrefixCls } from '@ikun-ui/utils';
	import { clsx } from 'clsx';
	import type { KAsideProps, KLayoutProps } from './types';

	export let cls: KLayoutProps['cls'] = undefined;
	export let attrs: KLayoutProps['attrs'] = {};

	export let width: KAsideProps['width'] = '';

	const prefixCls = getPrefixCls('aside');
	$: cnames = clsx(
		prefixCls,
		{
			[`${prefixCls}--base`]: true
		},
		cls
	);
</script>

<footer class={cnames} {...$$restProps} {...attrs} style:width>
	<slot />
</footer>
