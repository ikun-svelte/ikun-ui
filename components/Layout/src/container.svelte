<script lang="ts">
	import { getPrefixCls } from '@ikun-ui/utils';
	import { clsx } from 'clsx';
	import type { KContainerProps, KLayoutProps } from './types';

	export let cls: KLayoutProps['cls'] = undefined;
	export let attrs: KLayoutProps['attrs'] = {};
	export let direction: KContainerProps['direction'] = '';

	$: isVertical =
		direction === 'vertical' || (!direction && $$slots && ($$slots.header || $$slots.footer));

	const prefixCls = getPrefixCls('container');
	$: cnames = clsx(
		prefixCls,
		{
			[`${prefixCls}--base`]: true,
			[`${prefixCls}--base__vertical`]: isVertical
		},
		cls
	);
</script>

<section class={cnames} {...$$restProps} {...attrs}>
	<slot name="header" />
	<slot />
	<slot name="footer" />
</section>
