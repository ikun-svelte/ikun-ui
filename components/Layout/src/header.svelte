<script lang="ts">
	import { getPrefixCls } from '@ikun-ui/utils';
	import { clsx } from 'clsx';
	import type { KHeaderFooterProps, KLayoutProps } from './types';

	export let cls: KLayoutProps['cls'] = undefined;
	export let attrs: KLayoutProps['attrs'] = {};
	export let height: KHeaderFooterProps['height'] = '';
	const prefixCls = getPrefixCls('header');
	$: cnames = clsx(
		prefixCls,
		{
			[`${prefixCls}--base`]: true
		},
		cls
	);
</script>

<header class={cnames} {...$$restProps} {...attrs} style:height>
	<slot />
</header>
