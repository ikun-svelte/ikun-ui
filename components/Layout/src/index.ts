/// <reference types="./types" />
import Aside from './aside.svelte';
import Main from './main.svelte';
import Container from './container.svelte';
import Footer from './footer.svelte';
import Header from './header.svelte';
export {
	Aside as KAside,
	Main as KMain,
	Container as KContainer,
	Footer as KFooter,
	Header as KHeader
};

export type { KLayoutProps, KHeaderFooterProps, KAsideProps, KContainerProps } from './types';
