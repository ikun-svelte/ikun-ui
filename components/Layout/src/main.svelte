<script lang="ts">
	import { getPrefixCls } from '@ikun-ui/utils';
	import { clsx } from 'clsx';
	import type { KLayoutProps } from './types';

	export let cls: KLayoutProps['cls'] = undefined;
	export let attrs: KLayoutProps['attrs'] = {};

	const prefixCls = getPrefixCls('main');

	$: cnames = clsx(
		prefixCls,
		{
			[`${prefixCls}--base`]: true
		},
		cls
	);
</script>

<main class={cnames} {...$$restProps} {...attrs}>
	<slot />
</main>
