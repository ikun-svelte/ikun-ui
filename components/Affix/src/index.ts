/// <reference types="./types" />
import Affix from './index.svelte';
export { Affix as KAffix };

export default Affix;
