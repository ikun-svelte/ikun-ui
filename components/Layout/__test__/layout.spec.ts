import { afterEach, expect, test, describe, beforeEach } from 'vitest';
import { KMain, KAside, KHeader, KFooter } from '../src';
import KContainerSlot from './fixture/container.slot.svelte';
import KContainerDirection from './fixture/container.direction.svelte';
import { tick } from 'svelte';
let host: HTMLElement;

const initHost = () => {
	host = document.createElement('div');
	host.setAttribute('id', 'host');
	document.body.appendChild(host);
};
beforeEach(() => {
	initHost();
});
afterEach(() => {
	host.remove();
});

describe('Test: KMain', () => {
	test('KMain class render', () => {
		const instance = new KMain({
			target: host
		});

		expect(instance).toBeTruthy();
		expect(host.innerHTML.includes('k-main k-main--base')).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});
});

describe('Test: KAside', () => {
	test('props: width', () => {
		const instance = new KAside({
			target: host,
			props: {
				width: '666px'
			}
		});

		expect(instance).toBeTruthy();
		expect(host.innerHTML.includes('k-aside k-aside--base')).toBeTruthy();
		expect(host.innerHTML.includes('style="width: 666px;"')).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});
});

describe('Test: KHeader', () => {
	test('props: height', () => {
		const instance = new KHeader({
			target: host,
			props: {
				height: '666px'
			}
		});

		expect(instance).toBeTruthy();
		expect(host.innerHTML.includes('k-header k-header--base')).toBeTruthy();
		expect(host.innerHTML.includes('style="height: 666px;"')).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});
});

describe('Test: KFooter', () => {
	test('props: height', () => {
		const instance = new KFooter({
			target: host,
			props: {
				height: '666px'
			}
		});

		expect(instance).toBeTruthy();
		expect(host.innerHTML.includes('k-footer k-footer--base')).toBeTruthy();
		expect(host.innerHTML.includes('style="height: 666px;"')).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});
});

describe('Test: KContainer', () => {
	test('props: direction', async () => {
		const instance = new KContainerDirection({
			target: host,
			props: {
				direction: 'horizontal'
			}
		});
		expect(instance).toBeTruthy();
		expect(host.innerHTML.includes('k-header k-header--base')).toBeTruthy();
		expect(host.innerHTML.includes('k-main k-main--base')).toBeTruthy();
		expect(host.innerHTML.includes('k-footer k-footer--base')).toBeTruthy();
		expect(
			host.innerHTML.includes('k-container--base k-container--base__vertical')
		).not.toBeTruthy();
		instance.$set({ direction: 'vertical' });
		await tick();
		expect(host.innerHTML.includes('k-container--base k-container--base__vertical')).toBeTruthy();
	});

	test('slot: default & header & footer', async () => {
		const instance = new KContainerSlot({
			target: host
		});
		expect(instance).toBeTruthy();
		expect(host.innerHTML.includes('k-header k-header--base')).toBeTruthy();
		expect(host.innerHTML.includes('k-main k-main--base')).toBeTruthy();
		expect(host.innerHTML.includes('k-footer k-footer--base')).toBeTruthy();
		expect(host.innerHTML.includes('k-container--base k-container--base__vertical')).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});
});
