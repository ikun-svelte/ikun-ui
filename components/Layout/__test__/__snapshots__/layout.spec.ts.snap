// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Test: KAside > props: width 1`] = `"<footer class=\\"k-aside k-aside--base\\" style=\\"width: 666px;\\"></footer>"`;

exports[`Test: KContainer > slot: default & header & footer 1`] = `"<section class=\\"k-container k-container--base k-container--base__vertical\\"><header class=\\"k-header k-header--base\\" slot=\\"header\\">KHeader</header> <main class=\\"k-main k-main--base\\">KMain</main> <footer class=\\"k-footer k-footer--base\\" slot=\\"footer\\">KFooter</footer></section>"`;

exports[`Test: KFooter > props: height 1`] = `"<footer class=\\"k-footer k-footer--base\\" style=\\"height: 666px;\\"></footer>"`;

exports[`Test: KHeader > props: height 1`] = `"<header class=\\"k-header k-header--base\\" style=\\"height: 666px;\\"></header>"`;

exports[`Test: KMain > KMain class render 1`] = `"<main class=\\"k-main k-main--base\\"></main>"`;
